glean-core/csharp/Glean/Dispatchers.cs

```diff
@@ -3,24 +3,213 @@
+        /// function.  This includes all metric recording functions. For
+        /// launching of tasks that need to be processed asynchronously but
+        /// should not block other tasks, see `LaunchConcurrent`.
+        ///
+        /// If `QueueInitialTasks` is enabled, then the operation will be
+        /// created and added to the `taskQueue` but not executed until flushed.
+        ///
+        /// If `TestingMode` is enabled, then `LaunchAPI` will execute the task
+        /// immediately and synchronously to avoid asynchronous issues in tests.
+        /// </summary>
+        /// <param name="action">The Action to invoke</param>
         internal static void LaunchAPI(Action action)
         {
-            action();
+            if (QueueInitialTasks)
+            {
+                // If we are queuing tasks, typically before Glean has been
+                // initialized, then we should just add the created Task to the
+                // taskQueue.
+                AddTaskToQueue(new Task(action, cancellationToken));
+            }
+            else
+            {
+                if (!TestingMode)
+                {
+                    // If we are not queuing initial tasks, we can go ahead and
+                    // execute the task asynchronously on the ExclusiveScheduler
+                    new Task(action, cancellationToken)
+                        .Start(taskSchedulerPair.ExclusiveScheduler);
+                }
+                else
+                {
+                    // If we are in testing mode, then go ahead and invoke the
+                    // action to ensure synchronous execution.
+                    action.Invoke();
+                }
+            }
+         
         }
 
+        /// <summary>
+        /// This function launches an Action as an asynchronous Task on a
+        /// concurrently executed queue.
+        ///
+        /// **Note:** Tasks that need to be executed asynchronously but should
+        /// not block other tasks such as recording data should use the
+        /// `LaunchConcurrent` function. For example, tasks performed during
+        /// initialization or an upload task could be executed concurrently.
+        /// For tasks that need to be executed serially, see `LaunchAPI`.
+        ///
+        /// This function specifically ignores the `QueueInitialTasks` flag
+        /// because the only tasks that should be launched by this are the ping
+        /// upload schedulers and those should run regardless of the initialized
+        /// state.
+        ///
+        /// If `TestingMode` is enabled, then `LaunchConcurrent` will just
+        /// execute the task rather than adding it to the concurrent queue to
+        /// avoid asynchronous issues while testing.
+        /// </summary>
+        /// <param name="action">The Action to perform</param>
+        internal static void LaunchConcurrent(Action action)
```

Comment:
I find the `LaunchConcurrent` naming a bit confusing (I know that's how it works in Swift): isn't `LaunchAPI` concurrent as well?

The difference we have on the Kotlin side is this: `launch` adds the task at the end of the queue, while `execute` runs the task _as soon as possible_, right after the current executing task (if any). The executor is the same for both.

Reply:
In this case `LaunchAPI` uses a serially executed `ThreadPool` (single threaded, executes sequentially), while `LaunchConcurrent` uses a concurrently executed `ThreadPool` (multi-threaded, tasks run concurrently).  I was a little torn on implementing this, I originally started patterning it after the Kotlin implementation when I discovered the `ConcurrentExclusiveSchedulerPair` provided both types of execution in a single object, so I ended up going more towards the Swift implementation

Reply:
My concern is that this would end up working weirdly around startup: in kotlin, since we use the same pool of threads, init blocks the pool until after it finishes.

Reply:
These are separate ThreadPool's so I think it should be okay.  Let's see how the tests work out and go from there.